test/topology_experimental_raft/test_mv_tablets.py

```diff
@@ -73,3 +74,113 @@ async def test_tablet_mv_simple_6node(manager: ManagerClient):
     # We used SYNCHRONOUS_UPDATES=TRUE, so the view should be updated:
     assert [(3,2)] == list(await cql.run_async("SELECT * FROM test.tv WHERE c=3"))
     await cql.run_async("DROP KEYSPACE test")
+
+# Convenience function to open a connection to Alternator usable by the
+# AWS SDK. When more than one test needs it, we can move this function to
+# a separate library file. Until then, let's just have it in front of the
+# only test that needs it.
+import boto3
+import botocore
+alternator_config = {
+    'alternator_port': 8000,
+    'alternator_write_isolation': 'only_rmw_uses_lwt',
+}
+def get_alternator(ip):
+    url = f"http://{ip}:{alternator_config['alternator_port']}"
+    return boto3.resource('dynamodb', endpoint_url=url,
+        region_name='us-east-1',
+        aws_access_key_id='alternator',
+        aws_secret_access_key='secret_pass',
+        config=botocore.client.Config(
+            retries={"max_attempts": 0},
+            read_timeout=300)
+    )
+
+# Alternator convenience function for fetching the entire result set of a
+# query into an array of items.
+def full_query(table, ConsistentRead=True, **kwargs):
```

Comment:
wouldn't be better to import those functions from alternator/util? deduplication later is harder, especially in case the code diverges

Reply:
I would vote **no**. We should minimize the number of convenience functions, and it's not always possible to share them between suites because of completely different assumptions and coding styles. E.g., in this  Sometimes copying 7 lines of code is better than turning the world up-side-down just to be able to reuse a 7 line function.

Reply:
Duplicating the number of functions doesn't minimize it's number unless you count differently. My main point is that different suites should not be treated as separate projects, that's unfortunate. The coding style should converge instead of allowing more more differgence.

Reply:
1. I meant we should minimize the number of **global** convenience functions. I don't know how many dtests you wrote, but their global "convenience" functions are everything that I want to avoid:  I never understand them when I read a test, there are a hundred of them and each one takes a hundred different parameters. It's something I don't want to duplicate in my own tests.
2. I don't know if "suites should be treated as separate projects" but in reality they are. And to be honest, if I need to choose between duplicating a 7-line function and enforcing irrelevant standards across the board (e.g., you must use async in test/alternator because we wanted to use it in test/topology!), I prefer duplicating a 7-line function.

Specifically, I noted above in the comment that according to the test/topology_experimental_raft coding standard, these Alternator tests will eventually need to be moved to using some sort of async library, not boto3. This full_query implementation will need to become async as well. The one in Alternator won't. I don't want to make 800 tests in test/alternator harder to read and write just to make 7 tests in test/topology_experimental_raft happy.

Finally, my philosophy of these convenience functions is that they should be very short and very simple - the exact opposite of dtest's functions that take 10 different parameters to suite 10 different needs. Such short and simple functions should not "diverge" over time. They shouldn't even change (or change very rarely) - my whole point is that I don't want people to start adding more and more features to the existing functions.

Reply:
The problem you're describing falls more into `good library api design` category than duplicate vs code reuse. Comparing bad design to different bad design doesn't prove much in absolute terms. Agree that one should not create functions like
`def full_scan(table, ConsistentRead=True, Count=0, WithRestart=false, OnlyEven=true, **kwargs)`
but something like
`def full_scan(table)`
`def full_scan_and_count(table)`
but that's orthogonal to code duplication. It's not very bad though as you said function is very small. Just for the future I'd expect alternator part of topology tests to grow - ideally alternator should have some more system testing not only single node api tests.

I am not fully grasping async thing in python, mostly I think it's a distraction currently. I don't see why we should have both async and non async functions (see coroutines vs continuations pattern). Either this or that, mixed state should be seen as temporary not the goal.

Reply:
Let's reconsider the library issue when it's clearer what the library should contain. My experience with dtest's collection of convenience functions is terrible (I don't think that this review is the place to discuss it), whereas in cql-pytest it mostly worked because the "library" is so small, deliberately.

Regarding async - the topology guys decided to use it, but it's not used in test/cql-pytest. So I don't understand what you are proposing. Should we edit 800 Alternator test functions, almost 20,000 lines of code, to make everything async? Just so we could reuse a 7-line function (or maybe five such functions) between the projects? I'm not convinced.